### PR TITLE
proof !

### PR DESCRIPTION
**Pull request by jaljo**

### CODE REVIEW

File: text-editor/text-editor-002-en-proofing.md [hunk 1]
```diff
@@ -0,0 +1,1015 @@
+# How we built a medium like rich text editor
+
+This article synthethises the reflexion we went through during the conception
+and developpment of a web based rich text editor. We'll first focus on the
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: developpment -> development

File: text-editor/text-editor-002-en-proofing.md [hunk 2]
```diff
@@ -0,0 +1,1015 @@
...
+situation which led us to that developpment, then we'll explain the different
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: developpment -> development

File: text-editor/text-editor-002-en-proofing.md [hunk 3]
```diff
@@ -0,0 +1,1015 @@
...
+pathes we took to solve the problem and the one we finally chose.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: pathes -> paths

File: text-editor/text-editor-002-en-proofing.md [hunk 4]
```diff
@@ -0,0 +1,1015 @@
...
+
+See TL;DR at the end of the article for those of you who are speed up ;)
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: Tu as voulu dire "pour ceux d'entres vous qui sont rapides"? 
J'aurais plutôt mis "for those of you who are fast readers" ou quelque chose comme ça.

File: text-editor/text-editor-002-en-proofing.md [hunk 5]
```diff
@@ -0,0 +1,1015 @@
...
+    2. [Inserting the embed tweet into the DOM](#inserting-the-embed-tweet-into-the-DOM)
+    3. [Rendering the inserted tweet and going back to the original state](#rendering-the-inserted-tweet-and-going-back-to-the-original-state)
+9. [Clipboard access](#clipboard-access)
+10. [Save the edited content](#save-the-edited-content)
+11. [Conclusion](#conclusion)
+12. [TL;DR](#tldr)
+
+## Context
+
+We're working for an information company. They publish many contents on their
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: information company ->  a news company

File: text-editor/text-editor-002-en-proofing.md [hunk 6]
```diff
@@ -0,0 +1,1015 @@
...
+website, most of the time represented as articles or as a collection of
+articles (they call it `topics`). As the company is working in the information
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: information -> news

File: text-editor/text-editor-002-en-proofing.md [hunk 7]
```diff
@@ -0,0 +1,1015 @@
...
+business, they also post some `news`. News are small texts, which may contain
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: News are small texts -> News, in this case, are small texts

File: text-editor/text-editor-002-en-proofing.md [hunk 8]
```diff
@@ -0,0 +1,1015 @@
...
+tweets or links to other articles.
+
+Editors and web managers needed a tool to be able to create and edit these kinds
+of contents. These will be saved in our database as an HTML string.
+
+There alreay is many text edition tools online, but we wanted a tool which
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: There already is -> there already are

File: text-editor/text-editor-002-en-proofing.md [hunk 9]
```diff
@@ -0,0 +1,1015 @@
...
+looks close to the modern text editor of [medium.com](https://medium.com/).
+Some open source libraries are offering such solution, but most of the time,
+when specific needs arise, they no longer fulfill all your
+requirements. And here you are on the highway of awfull tweaks... Instead of
+spending time to develop additional features on top of an existing library (with
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: spending time to develop -> spending time developing

File: text-editor/text-editor-002-en-proofing.md [hunk 10]
```diff
@@ -0,0 +1,1015 @@
...
+all drawbacks implied), we decided to implement our own tool. That way, we could
+make it behave exactly the way we want.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: we want -> we wanted

File: text-editor/text-editor-002-en-proofing.md [hunk 11]
```diff
@@ -0,0 +1,1015 @@
...
+
+Keep in mind that our text editor development took place whereas the frontend
+website used to publish informations was close to be put online.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: informations -> information

File: text-editor/text-editor-002-en-proofing.md [hunk 12]
```diff
@@ -0,0 +1,1015 @@
...
+
+Technically speaking, articles are fetched from an HTTP REST API (our back
+office). Some of them were coming from a legacy application, and may contain
+some broken HTML semantic. To address this problem, we decided to develop a
+parser for the received HTML markup to be cleaned. This parser creates React
+components for, and only for HTML tags that we allow. Doing so, only tags we're
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: for, and only for -> for and only for

File: text-editor/text-editor-002-en-proofing.md [hunk 13]
```diff
@@ -0,0 +1,1015 @@
...
+sure to support are rendered, which also simplifies design and stylization.
+
+The same parsing logic had to be kept for our text editor, as we wanted it to be
+WYSIWYG: the edited content is displayed like it would be in the public website,
+with the ability to directly edit it.
+
+## The specs
+
+The text editor sould provide the basic fatures of text edition :
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: sould -> should , fatures -> features

File: text-editor/text-editor-002-en-proofing.md [hunk 14]
```diff
@@ -0,0 +1,1015 @@
...
+- create titles
+- create quotes
+
+Our client also needed to insert some more specific medias :
+- regular images (uploaded or picked from their image collection)
+- videos (also picked from their video collection)
+- tweets (by pasting the tweet URL)
+- Youtube videos (by pasting the video URL)
+
+Although such features may seems very common, complexity shows up because our
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: may seems -> may seem

File: text-editor/text-editor-002-en-proofing.md [hunk 15]
```diff
@@ -0,0 +1,1015 @@
...
+text editor fits in a more global administration system. Unlike in a regular
+text editor, where an user can insert images browsed from it's computer, we had
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: an user -> a user (c'est une exception, car ça se prononce "yi" donc pas de -n après le a :) ) ; it's computer -> their computer

File: text-editor/text-editor-002-en-proofing.md [hunk 16]
```diff
@@ -0,0 +1,1015 @@
...
+```js
+{
+    type: "text",
+    text: "lorem ipsum",
+    children: [],
+}
+```
+[See gist here](https://gist.github.com/jaljo/b07574e0dff05acaa996a207d2a68fd3)
+
+was succesfully transformed as a child of an other node when trying to make it
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: an other -> another

File: text-editor/text-editor-002-en-proofing.md [hunk 17]
```diff
@@ -0,0 +1,1015 @@
...
+            children: []
+        }
+    ]
+}
+```
+[See gist here](https://gist.github.com/jaljo/bedbb22286709ff6c8a895bd0c742a5e)
+
+However, things were getting way more complicated from the moment we worked on
+more edgy cases, i.e. update some text that was starting in a paragraph and
+ending in an other. Selection overlapsing on different and nested tags, such as
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: an other -> another

File: text-editor/text-editor-002-en-proofing.md [hunk 18]
```diff
@@ -0,0 +1,1015 @@
...
+
+```
+<b>Hello <em>John Doe</em></b>, how are you ?
+    |---------------|
+selection start   selection end
+```
+[See gist here](https://gist.github.com/jaljo/0408185bff348ba189fd79aec1984931)
+
+was a daunting too. Applying such changes on the AST would have been difficult
+for us. In the mean time, the developer who started to work on the AST was
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: mean time -> meantime (sinon c'est que le temps est méchant ;) )

File: text-editor/text-editor-002-en-proofing.md [hunk 19]
```diff
@@ -0,0 +1,1015 @@
...
+required on an other project, so we decided not to use this AST anymore. Because
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: an other -> another

File: text-editor/text-editor-002-en-proofing.md [hunk 20]
```diff
@@ -0,0 +1,1015 @@
...
+it was the corner stone of the back office application, keeping it at that level
+of complexity and abstraction was too risky and time consuming to develop also
+to maintain, plus the learning curve was giving nightmares to newcomers.
+The whole project beeing based on functional programing principles, AST
+development should have embraced these principles as well. Thing is, AST
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: Thing is -> The thing is

File: text-editor/text-editor-002-en-proofing.md [hunk 21]
```diff
@@ -0,0 +1,1015 @@
...
+manipulation relies mostly on one ability to identifiy ancestors of a node,
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: identifiy -> identify

File: text-editor/text-editor-002-en-proofing.md [hunk 22]
```diff
@@ -0,0 +1,1015 @@
...
+find, update or even remove it down through a collection of other nodes.
+This implies a lot of recursion and make immutabilty very hard to preserve.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: make -> makes

File: text-editor/text-editor-002-en-proofing.md [hunk 23]
```diff
@@ -0,0 +1,1015 @@
...
+Although it would have been possible to achieve such result using advanced FP
+concepts like algebraic data types, we felt that it was definitely not a good
+bet to take for the project's sake.
+
+We took an other approach to the problem, by using the DOM API to keep the tree
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: an other -> another

File: text-editor/text-editor-002-en-proofing.md [hunk 24]
```diff
@@ -0,0 +1,1015 @@
...
+([gist](https://gist.github.com/jaljo/975cc2d57a8a184b49a7fc5caf94cc30))
+
+That sanitization part was performed once the new `b` tag was added to the DOM.
+
+This was working fine, but the number of lines of code started to dangerously
+increase and we were discovering more and more edge cases, which were a PITA to
+cover.
+
+Finally, down the corner of a murky russian Stackoverflow thread we found it.
+The method which would have saved us all.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: The method which would have saved us all -> The method which saved us all
Sinon ça ferait "la méthode qui nous aurait sauvé" (enfin il me semble)

File: text-editor/text-editor-002-en-proofing.md [hunk 25]
```diff
@@ -0,0 +1,1015 @@
...
+([gist](https://gist.github.com/jaljo/ae56cab294feb22d393675dedda48de3))
+
+and that's it ! Your selection is wrapped inside a `b` tag, the produced
+markup is very clean (remember the previous step where we had to sanitize the
+markup manually).
+
+This is very neat, it enables rich text edition directly in the browser (and
+by the browser :p). And look at that compatibility table, green everywhere !
+Icing on the cake: because we're working on `contentEditable` tags, conventional
+shortucts are working as well (CTRL+B for bold, etc).
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: shortucts -> shortcuts

File: text-editor/text-editor-002-en-proofing.md [hunk 26]
```diff
@@ -0,0 +1,1015 @@
...
+We're using
+- [react](https://reactjs.org/)
+- [redux](https://react-redux.js.org/)
+- [react-functional-lifecycle](https://github.com/Aloompa/react-functional-lifecycle)
+- [redux-observable](https://redux-observable.js.org/)
+- [ramda](https://ramdajs.com/)
+
+In the following part of this article, we'll assume that you're comfortable with
+these technologies. If you're not, please take some time to understand the core
+principles of them.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: the core principles of them -> their core principles

File: text-editor/text-editor-002-en-proofing.md [hunk 27]
```diff
@@ -0,0 +1,1015 @@
...
+          key={index}
+        />
+      )(componentInfos)
+  }
+</TextEditor>
+```
+[See gist here](https://gist.github.com/jaljo/9f8e07140c334f44d4a944008827e405)
+
+The `TextEditor` is wrapping a `Widget` component tree, which is the component
+responsible to render the HTML tags we support (remember what's explained in the
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: responsible to render -> responsible for rendering

File: text-editor/text-editor-002-en-proofing.md [hunk 28]
```diff
@@ -0,0 +1,1015 @@
...
+[context](#context) section). The resulting sanitized components are those which
+will be edited.
+
+The `main` property indicates if this instance of the TextEditor is the main
+instance used on the page. We're using this prop to filter some actions that
+only needs to be performed over that main text editor (see the container
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: Ce sont les actions qui ont besoin d'être "performed over"? Si oui -> "only need", mais si c'est "this prop", c'est bien "only needs"

File: text-editor/text-editor-002-en-proofing.md [hunk 29]
```diff
@@ -0,0 +1,1015 @@
...
+- `willUnmount`: dispatches the `CLEAR` action for the main text editor when
+it will be unmounted. This action is observed by the same epic to unregister the
+`window.mousedown` event listener.
+
+### Epics
+
+Our epics are exposed with
+[redux-observable](https://redux-observable.js.org/docs/basics/Epics.html).
+
+All the DOM manipulations and media insertion logic is done here. Notice that
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: is done -> are done

File: text-editor/text-editor-002-en-proofing.md [hunk 30]
```diff
@@ -0,0 +1,1015 @@
...
+
+## Text toolbox workflow
+
+The text toolbox contains a set of buttons to apply text mutations over the
+selected text :
+
+![text toolbox](/images/text-editor/text-toolbox.png)
+*The text toolbox*
+
+The component's view basically consist into a list of buttons :
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: consist into -> consists of

File: text-editor/text-editor-002-en-proofing.md [hunk 31]
```diff
@@ -0,0 +1,1015 @@
...
+  isQuote: false,
+  isLink: false,
+}
+```
+[See gist here](https://gist.github.com/jaljo/43baf55906136340507325ba39938494)
+
+The `isTitle` flag is used to disable the  bold / unbold button when the
+selected text is a title (we always want titles to be bold).
+The `isBold` flag is used to whether highlight or not the bold / unbold button
+when the selected text is aleady bold or not (this button behavior is similar
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: aleady -> already

File: text-editor/text-editor-002-en-proofing.md [hunk 32]
```diff
@@ -0,0 +1,1015 @@
...
+to the one you'll find on any wording processor).
+When clicked, the button dispatches a `MUTATE` action. That action is observed
+by an epic which will actually perform the selected text mutation.
+
+**How these flags values are determined ?**
+A dedicated epic is executed both when we open the text toolbox, and also when
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: Si tu utilises "both", pas besoin d'ajouter ''also" si c'est pour la même phrase : "both when we open the text toolbox and apply a text mutation". Mais si tu préfères ta tournure de phrase, tu peux juste retirer "both" mais bien mettre "when we apply" et non "applying". :)

File: text-editor/text-editor-002-en-proofing.md [hunk 33]
```diff
@@ -0,0 +1,1015 @@
...
+      isQuote: isInParent('blockquote')(range),
+      isLink: isInParent('a')(range),
+    })]),
+    map(apply(refreshButtonsState)),
+    logObservableError(),
+  )
+```
+[See gist here](https://gist.github.com/jaljo/97a4648a42a9ca2b81ecfc313d81d837)
+
+**How the text is mutated ?**
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: How the text is mutated -> How is the text mutated ? (tu peux mettre ta phrase, mais sans le "?")

File: text-editor/text-editor-002-en-proofing.md [hunk 34]
```diff
@@ -0,0 +1,1015 @@
...
+{ /* TextEditor component's view */ }
+<article
+  onSelect={() => selectText(editorName)}
+>
+  {children}
+</article>
+```
+[See gist here](https://gist.github.com/jaljo/ae7cb56eb3aa526e5ea66eed1c0f260c)
+
+This action is also dispatched when moving the cursor, so it make it easy to
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: it make -> it makes

File: text-editor/text-editor-002-en-proofing.md [hunk 35]
```diff
@@ -0,0 +1,1015 @@
...
+      getParagraphTopPosition(selection),
+      getNodeIndex(editor, selection.anchorNode),
+    ]),
+    map(apply(showParagraphToolbox)),
+    logObservableError(),
+  )
+```
+[See gist here](https://gist.github.com/jaljo/2a43d542d80077b6dda253f7f5a107b1)
+
+This epic show the paragraph toolbox for, and only for the editor having the
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: This epic show -> This epic shows
for, and only for -> to and only to
and only for the editor having the right... -> and only to the editor who has

File: text-editor/text-editor-002-en-proofing.md [hunk 36]
```diff
@@ -0,0 +1,1015 @@
...
+right `editorName`. The absolute position at which the paragraph toolbox will be
+shown is computed from there. We also store in the redux state the node index
+corresponding to the paragraph in which the cursor is. It will later be used
+to insert medias chosen with the paragraph toolbox.
+
+Unlike the text toolbox buttons, paragraph toolbox buttons each behave
+differently. Explaining them all will be too long, we'll rather focus on an
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: on an speaking example -> on a telling example

File: text-editor/text-editor-002-en-proofing.md [hunk 37]
```diff
@@ -0,0 +1,1015 @@
...
+speaking example. For other cases, feel free to check the source code !
+
+## Insertion of a tweet : a case study
+
+In this chapter, we will try to give a better insight of how the whole thing is
+working, by breaking down a simple use case and the components involved in it's
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: it's workflow -> its workflow

File: text-editor/text-editor-002-en-proofing.md [hunk 38]
```diff
@@ -0,0 +1,1015 @@
...
+
+Keep in mind that our main goal here is to obtain the cleanest possible HTML
+representation of our tweet, so it can lately be saved to the API. That way it
+will be easy for us to parse it as a component, and finally display it to the
+end user in the front end application.
+
+The following chart will give you an overview of the actions flow :
+![Insert tweet action flow chart](/images/text-editor/RTE.png)
+
+History begins as soon as an user submits the little tweet insertion form.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: an user -> a user

File: text-editor/text-editor-002-en-proofing.md [hunk 39]
```diff
@@ -0,0 +1,1015 @@
...
+Thanks to `redux` action creators, the url of the tweet he wants to insert is
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: the tweet he wants to -> the tweet they want to (c'est un "singular they" car on ne sait pas si c'est un "he" ou "she")

File: text-editor/text-editor-002-en-proofing.md [hunk 40]
```diff
@@ -0,0 +1,1015 @@
...
+the end of this article.
+
+The first epic is only responsible to call that endpoint and dispatch a
+`EMBED_TWEET_FETCHED` action which carries out the obtained HTML markup in its
+payload. It's always convenient to isolate API calls in dedicated epics so
+errors can be properly catched and displayed to the end user when something goes
+wrong.
+Cool thing is, keeping epics as atomic as possible (observe an action, do one
+thing, then dispatch another action) allow them to be easily tested. That is,
+your fellow developpers dont want to break your legs when the prod crashes
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: developpers -> developers 
dont -> don't

File: text-editor/text-editor-002-en-proofing.md [hunk 41]
```diff
@@ -0,0 +1,1015 @@
...
+(which will happend, but that's another story).
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: which will happend -> which will happen

File: text-editor/text-editor-002-en-proofing.md [hunk 42]
```diff
@@ -0,0 +1,1015 @@
...
+)
+
+// getEditor :: String -> Node
+export const getEditor = editorName => document.querySelector(
+  `.edited-text-root[data-editor-name="${editorName}"]`
+)
+```
+[See gist here](https://gist.github.com/jaljo/4c8f83acc48766b930d7d584fe9ed22b)
+
+As you can see, the `insertTweetNode` function returns a Promise, it provide
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: provide -> provides

File: text-editor/text-editor-002-en-proofing.md [hunk 43]
```diff
@@ -0,0 +1,1015 @@
...
+us a nice way to handle failures at that step of the process. What does that
+function do ?
+
+- Using a combination of react `renderToString` utility, one of our parsers, and
+the HTML markup provided by the twitter API in the previous observable, it
+creates an HTML representation of the tweet **formatted as we need it to be.**
+- Then, using an `insertNewNodeAtIndex` utility, we replace the empty paragraph
+from where the tweet insertion form was shown by the new node created at the
+previous step (that paragraph index has early been saved into the paragraph
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: early -> already

File: text-editor/text-editor-002-en-proofing.md [hunk 44]
```diff
@@ -0,0 +1,1015 @@
...
+dispatched by the epic. Otherwise, an error is catched and displayed to the
+user. Notice that the kind of error we handle here is not of the same type than
+in the previous observable. Errors that may occur here could only be DOM
+related.
+
+### Rendering the inserted tweet and going back to the original state
+
+The `TWEET_INSERTED` action is observed by four epics (and here we gain full
+benefit of working with observable streams) :
+- One that takes care of systematically insert and focus a new paragraph after a
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: One that takes care of systematically insert and focus a new paragraph after a -> One that systematically takes care of inserting and focusing a new paragraph after a

File: text-editor/text-editor-002-en-proofing.md [hunk 45]
```diff
@@ -0,0 +1,1015 @@
...
+media has been inserted (for pure UX concerns),
+- One that will hide the paragraph toolbox (we no longer need it to be displayed
+as a consequence of the previous action: a new paragraph has been inserted),
+- One that will close the tweet insertion form so that the user is not annoyed
+when the paragraph toolbox is lately shown,
+- The last wich only maps the payload of the `TWEET_INSERTED` action payload to
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: The last wich -> The last which

File: text-editor/text-editor-002-en-proofing.md [hunk 46]
```diff
@@ -0,0 +1,1015 @@
...
+a new `RENDER_TWEET` one. Although it can sound a bit overkill to have an epic
+that does apparently "nothing", it help us keep our action stream crystal clear.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: it help -> it helps

File: text-editor/text-editor-002-en-proofing.md [hunk 47]
```diff
@@ -0,0 +1,1015 @@
...
+      originalHtmlMarkup,
+      window.twttr.widgets.createTweet(tweetId, element),
+    ])),
+    map(([ tweetId, originalHtmlMarkup ]) => tweetRendered(tweetId, originalHtmlMarkup)),
+    logObservableError(),
+  )
+```
+[See gist here](https://gist.github.com/jaljo/006cab23716b9076dfcd5b2032d03081)
+
+That epic check for the twitter SDK presence in the window object using the
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: That epic check -> That epic checks

File: text-editor/text-editor-002-en-proofing.md [hunk 48]
```diff
@@ -0,0 +1,1015 @@
...
+`filter` operator. Noticed how we pass the window object as a dependency of
+the epic ? This life saving feature is given by rxjs 6 and help us to keep our
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: and help -> and helps

File: text-editor/text-editor-002-en-proofing.md [hunk 49]
```diff
@@ -0,0 +1,1015 @@
...
+
+- The tweet widget is created by a call to the Twitter SDK, which relies on the
+`tweetId` extracted from the URL and an HTML element. Those two elements are
+derived of the original action payload (remember how we map them in the resolved
+promise of the `insertTweetNode` function).
+- Once this has be done, we dispatch the last (phew !) `TWEET_RENDERED` action
+to resolve our observable stream.
+
+That final action is reduced so the rendered tweet markup is appened to a
+collection of others rendered tweets.
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: others rendered tweets -> other rendered tweets

File: text-editor/text-editor-002-en-proofing.md [hunk 50]
```diff
@@ -0,0 +1,1015 @@
...
+
+## Clipboard access
+
+Using `contentEditable` tags rather than regular form elements such as `input`
+has drawbacks. Say that our end users need to paste text from other sources to
+write articles. Doing so in an input element, everything works just great. You
+get rid of non textual elements, and end up with a raw string ready to be
+submitted. When pasting in a `contentEditable` element, things are way
+different.
+The browser will try to render the pasted text as close as possible from of the
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: from of the -> from the

File: text-editor/text-editor-002-en-proofing.md [hunk 51]
```diff
@@ -0,0 +1,1015 @@
...
+original copied text, resulting in massive and ugly inline style everywhere.
+This is not what we want, because we tried to have a clean HTML semantic by
+controlling **how elements are rendered** (remember we wrote a parser for that).
+
+We had to find a way to sanitize (once more !) the user clipboard, to ensure
+only raw text is pasted from external sources. Problem is, clipboard interaction
+is one of the most obvious security breach in web application, because it
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: in web application -> in web applications

File: text-editor/text-editor-002-en-proofing.md [hunk 52]
```diff
@@ -0,0 +1,1015 @@
...
+implies a direct communication between javascript and the external world with
+almost no way to control what's comming from it. In the past, several workaround
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: what's comming -> what's coming
several workaround -> several workarounds

File: text-editor/text-editor-002-en-proofing.md [hunk 53]
```diff
@@ -0,0 +1,1015 @@
...
+involving extreme methods such as `iframe` and even `flash` (brrrr) were used.
+Lately, the well known `document.execCommand('paste')` came to the rescue, but
+was still very limitated due to security concerns. See
+[that link](https://frontendnews.io/editions/2018-08-01-copy-and-paste-clipboard-api)
+for details.
+
+A more recent approach was to use the
+[new clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API),
+which provide an elegant and secured way to access the user clipboard (as soon
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: which provide -> which provides

File: text-editor/text-editor-002-en-proofing.md [hunk 54]
```diff
@@ -0,0 +1,1015 @@
...
+as he gaves the app the permission to do so) and let us work with Promises in
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: let us work -> allows us to work

File: text-editor/text-editor-002-en-proofing.md [hunk 55]
```diff
@@ -0,0 +1,1015 @@
...
+return. This feature is still at experimental state and the API is likely to
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: at experimental state -> at its experimental state

File: text-editor/text-editor-002-en-proofing.md [hunk 56]
```diff
@@ -0,0 +1,1015 @@
...
+evolve in the future. We use it anyway, because it's what fits our user need the
+best, and we were lucky enough for the compatibility tables to match his
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: to match his -> to match its

File: text-editor/text-editor-002-en-proofing.md [hunk 57]
```diff
@@ -0,0 +1,1015 @@
...
+environment.
+
+If you recall, an `onPaste` handler was binded to the article editor. It is
+triggered either when user pastes text from the shortcut key combination CTRL+V
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: user pastes -> soit "when users paste" soit "when the user pastes"

File: text-editor/text-editor-002-en-proofing.md [hunk 58]
```diff
@@ -0,0 +1,1015 @@
...
+    isClipboardAccessGranted,
+    () => navigator.clipboard.readText().then(pasteGranted),
+    () => [displayClipboardWarning()],
+  )),
+  logObservableErrorAndTriggerAction(displayClipboardSupportError),
+)
+```
+[See gist here](https://gist.github.com/jaljo/4e22647075c2199e52a4eb93dad0f672)
+
+This first epic check for the clipboard access and dispatches an error message
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: This first epic check -> The first epic checks

File: text-editor/text-editor-002-en-proofing.md [hunk 59]
```diff
@@ -0,0 +1,1015 @@
...
+The rendered representations are always more verbose than the markup code,
+that's why we prefer to store only the embed markup in our DB.
+Additionaly, the legacy text edition application was already working with
+embed code to display tweets, so we wanted to make this TextEditor
+compatible with the contents previously edited with that legacy app.
+
+This brings us to the `sanitization` step, where we replace any rendered
+representation of complex inserted media by their embed markup. To do so,
+we have to keep the embed markup of the media in the application state.
+We also have to identify to which embed markup is correponding the rendered
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: is correponding -> is corresponding

File: text-editor/text-editor-002-en-proofing.md [hunk 60]
```diff
@@ -0,0 +1,1015 @@
...
+
+Doing so, we're sure to persist a valid string in the database :)
+
+The saved string can then be used by the public website to display the article.
+
+## Conclusion
+
+Phew ! That was a long ride ! Congratulations for staying with us so far ;)
+At that time of writing, the text editor is still being improved to follow
+feedbacks given by end users. We described it as it is when we wrote this
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: as it is when -> as it was when

File: text-editor/text-editor-002-en-proofing.md [hunk 61]
```diff
@@ -0,0 +1,1015 @@
...
+article.
+A possible improvement would be the ability to move some selected text by
+dragging and dropping it at the desired place. This could lead to an other blog
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: an other -> another

File: text-editor/text-editor-002-en-proofing.md [hunk 62]
```diff
@@ -0,0 +1,1015 @@
...
+post.
+
+It was a good excercise to develop. In the end, the concepts are most of all
+very simple. We learned a lot on the native browser functionalities
+and we are amazed by the possibilities of these sofwares !
```
Comment (Amandine3003) [inline text-editor/text-editor-002-en-proofing.md]: sofwares -> softwares

**Review by Amandine3003 — CHANGES_REQUESTED** [review]
Petits typos :)

File: text-editor/text-editor-002-en-proofing.md (hunk 4 above)
Comment (nicolasmure) [reply text-editor/text-editor-002-en-proofing.md]: Actually, it's not that they are fast readers, but it's more that they don't want to spend time / don't have a lot of time to read the article ^^.
So maybe we can change it for `for those of you who are in a hurry` ?

Comment (jaljo) [reply text-editor/text-editor-002-en-proofing.md]: J'ai voulu dire "pour ceux qui sont pressés", du coup j'ai remplacé par "those of you who are in a hurry" :)

**Comment by nicolasmure** [comment]
Thank you for the review @Amandine3003 !! :heart:

File: text-editor/text-editor-002-en-proofing.md (hunk 18 above)
Comment (jaljo) [reply text-editor/text-editor-002-en-proofing.md]: :D

File: text-editor/text-editor-002-en-proofing.md (hunk 32 above)
Comment (jaljo) [reply text-editor/text-editor-002-en-proofing.md]: La tienne est plus légère :)

File: text-editor/text-editor-002-en-proofing.md (hunk 35 above)
Comment (jaljo) [reply text-editor/text-editor-002-en-proofing.md]: Dans ce cas "the editor" designe notre editeur de texte et pas une perosonne, donc c'est bien for qu'il faut mettre non ?

**Comment by jaljo** [comment]
I fixed the typos, thank you so much :heart:

**Pull request merged**